server-protocol.md

```diff
@@ -186,16 +186,26 @@ observers to use the hard-earned token for themselves.
 Wormhole codes look like `4-purple-sausages`, consisting of a number followed
 by some random words. This number is called a "Nameplate".
 
-On the Rendezvous Server, the Nameplate contains a pointer to a Mailbox.
-Clients can "claim" a nameplate, and then later "release" it. Each claim is
+On the Mailbox server, the Nameplate contains a pointer to a Mailbox.
+Clients can "claim" a nameplate, each claim is
 for a specific side (so one client claiming the same nameplate multiple times
-only counts as one claim). Nameplates are deleted once the last client has
-released it, or after some period of inactivity.
+only counts as one claim). There used to be an explicit `release` command, but
+Nameplates are new implicitly deleted once both clients have claimed
+it and thus found their mailbox, or after some period of inactivity. A client
```

Comment:
Do clients still learn explicitly when they definitely have (or don't) a nameplate?

Reply:
How do you mean?

Reply:
When a client connects and has no chance of ever finding a peer (as is the case in the described scenarios), the best thing is to give it an error message when it tries to `claim` the nameplate. The alternative is to just free the nameplate and let the client go stale, which is the status quo AFAICT.

Reply:
Yes, I think proposing something to fix that hole could be good .. what I'm getting at here, though, is that the above words change the protocol from "explicit" (you must `RELEASE` the nameplate) to "implicit" so we need to make sure the clients still know the status of their nameplate.

Reply:
You mean that old clients may expect the nameplate to exist until they explicitly send "release"? That's a valid concern, however I don't see how this information could make any change in behavior that may be relevant here. Because once they've connected to the mailbox, they don't care about the nameplate anymore anyways.

Do you have some scenario in mind where my idea might go wrong?

Reply:
If a client thinks it still "has" a nameplate, and gets disconnected in that state, it will reconnect and try to "claim". I haven't analyzed the above to learn if that _could_ happen here though.

Regardless, it's often better "in general" to be explicit and consistent, is why I'm pursing this angle.

Reply:
The problem with "explicit" here is that we must wait until clients release it, which they currently only do until after the encryption handshake for some reason, which leaves us open to situations like in https://github.com/magic-wormhole/magic-wormhole-mailbox-server/issues/19

As far as I understand, a client, once ever claimed a nameplate, should always reconnect with `open` only. Should this not be the case in the Python implementation, then I'll change the wording and server side state machine to track this so that it cannot go wrong in the way you described, while still preventing the race hazard I'm trying to fix.

Reply:
> then I'll change the wording and server side state machine to track this so that it cannot go wrong in the way you described, while still preventing the race hazard I'm trying to fix.

After three hours of trying, it turns out that this is a lot more difficult than it should be. The problem is that after opening a mailbox, clients don't know who else is there or not until messages have been sent.

So either we come to the conclusion that your concern here is a non-issue and keep the current proposal, or I'll remove that bit and try to tackle it again once we do proper session closing.